joomla/joomla-cms
# Add Caching option to ModuleHelper to allow cached module content for logged in users

Pull Request for New feature

### Summary of Changes
Joomla doesn't use the cache when a user is logged in, it only uses cache for visitors (public / guest).
When you have a (membership) site where users normally are logged in and have alot of articles and 'expensive / high-load / remote content' modules, you will find that the server performance decreases a lot: the load generated for NOT being able to use cached versions of the modules will have a high impact.

Use Case: I have build a module for a customer that does several remote API calls to fetch stock information. Depending on the response times for this api, fetching and rendering the data can take up to 20 seconds. Visitors get the module content instantly because the content is cached, but as soon as a user logs in, the module content doesn't come from cache but needs to be rendered. When you have a lot of logged in users, this will have a high performance impact on the server.

This PR adds a cache_loggedin parameter (default set to 0 = off) that when enabled will also serve the module from cache for logged in users, or when no cache available will add the cache for it.

This PR adds the cache_loggedin parameter to all mod_article-xyz modules

### Testing Instructions
1. install PR
2. ensure that caching for the website is enabled
3. add a module mod_articles_latest to your website displaying random articles (order: Random Articles)
4. visit the front-end as guest and press several times ctrl-f5: the module will **show the same** articles for each ctrl - f5
5. login to the front-end
6. press several times ctrl-f5, the module will **show different** articles for each ctrl-f5

7. in the back-end change in the mod-article_latest, advanced tab the Cache logged In setting to 'Yes'
8. in the front-end (logged in), press several times ctrl - f5: the module will refresh and every next ctrl - f5 will **show the same** articles (because the module content now comes from cache)

You can also verify that when logged in on the front-end and the Cache Logged In setting in the module is set to No, there will be no cache file created for mod_articles_latest, when the Cache Logged In setting in the module is enabled a cache file for mod_latest_articles will be created.



### Expected result
see testing instructions


### Actual result
see testing instructions


### Documentation Changes Required
?
